src/Security/Authentication/Certificate/src/CertificateValidationCache.cs

```diff
@@ -0,0 +1,39 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Security.Cryptography.X509Certificates;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Caching.Memory;
+
+namespace Microsoft.AspNetCore.Authentication.Certificate
+{
+    /// <summary>
+    /// MemoryCache based implementation used to store <see cref="AuthenticateResult"/> results after the certificate has been validated
+    /// </summary>
+    public class CertificateValidationCache : ICertificateValidationCache
+    {
+        private MemoryCache _cache = new MemoryCache(new MemoryCacheOptions { SizeLimit = 1024 });
+
+        /// <summary>
+        /// Get the <see cref="AuthenticateResult"/> for the connection and certificate.
+        /// </summary>
+        /// <param name="connection">The connection.</param>
+        /// <param name="certificate">The certificate.</param>
+        /// <returns>the <see cref="AuthenticateResult"/></returns>
+        public AuthenticateResult Get(ConnectionInfo connection, X509Certificate certificate)
+            => _cache.Get(ComputeKey(connection, certificate)) as AuthenticateResult;
+
+        /// <summary>
+        /// Store a <see cref="AuthenticateResult"/> for the connection and certificate
+        /// </summary>
+        /// <param name="connection">The connection.</param>
+        /// <param name="certificate">The certificate.</param>
+        /// <param name="result">the <see cref="AuthenticateResult"/></param>
+        public void Put(ConnectionInfo connection, X509Certificate certificate, AuthenticateResult result)
+            =>  _cache.Set(ComputeKey(connection, certificate), result, new MemoryCacheEntryOptions().SetSize(1));
+
+        private string ComputeKey(ConnectionInfo connection, X509Certificate certificate)
+            => $"{certificate.GetCertHashString()}:{connection.Id}";
```

Comment:
@blowdart is the cert hash string a strong enough identifier? Or is the connectionId needed too?

It would be even more performant if we could cache the cert validation across connections, but that risks collisions.

Reply:
No. It's SHA1. Compute a SHA256 over the RawData property (and maybe make that an extension method that lives in X509CertificateExtensions.cs for neatness?

If SHA256 has collisions we have bigger problems :)

Reply:
We can just use GetCertHashString(HashAlgorithmName.SHA256), we don't need the connection id then right?